modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -450,13 +451,35 @@ public Hit getHitInfo(float x, float y) {
                 int[] trailing = new int[1];
                 charIndex = run.getStart() + run.getOffsetAtX(x, trailing);
                 leading = (trailing[0] == 0);
+
+                insertionIndex = charIndex;
+                if (getText() != null) {
+                    if (!leading) {
+                        BreakIterator charIterator = BreakIterator.getCharacterInstance();
+                        charIterator.setText(new String(getText()));
+                        int next = charIterator.following(insertionIndex);
+                        if (next == BreakIterator.DONE) {
+                            insertionIndex += 1;
+                        } else {
+                            insertionIndex = next;
+                        }
+                    }
+                } else if (!leading) {
+                    insertionIndex += 1;
+                }
             } else {
                 //empty line, set to line break leading
                 charIndex = line.getStart();
                 leading = true;
             }
         }
-        return new Hit(charIndex, -1, leading);
+        if (insertionIndex == -1) {
+            insertionIndex = charIndex;
+            if (!leading) {
+                insertionIndex += 1;
```

Comment:
I would have initialized insertionIndex directly in all the blocks instead of adding conditional logic, but this code does seem to produce the desired result. 
For clarify, would it be possible to avoid this additional conditional logic?

Also, do existing unit tests cover the newly modified paths?

Reply:
Initializing the insertion index directly with character index looks better. In this way we can avoid additional conditional logic as well. Changed code to do so.
Added additional test case. This tests the condition present in line 430.
I couldn't create the scenario to hit line 473. If you have any suggestions please let me know.

While testing I found that `testTextFlowInsertionIndexUsingMultipleEmojis()` was failing intermittently when all the tests are run. Hence added small delay.

Reply:
> I couldn't create the scenario to hit line 473. If you have any suggestions please let me know.

I can't seem to trigger this code path (I wonder if it's possible to hit it at all).  Anyway, the code there looks correct to me, even if this condition would never happen.

Reply:
Assuming line 473 is the one still there today, it looks to me as if that would be reached if you had the caret on an empty line that isn't the last line.

Reply:
Regarding
"If we can initialize insertionIndex to a positive value, then the buggy code in HitInfo.getInsertionIndex()
will never get executed, and we can remove it later (in JDK-8302511)."

it seems to me that this new code has been pretty much copied from the supposedly buggy code there .. 
I'm not sure that code is actually buggy (comments over in that bug), rather that the instance was constructed
with bad data.

Reply:
The main issue is that insertion index was not computed in the PrismTextLayout, supposedly to allow for delaying the computation until it's needed (if needed at all).  A good intention, but the execution had two issues:
1. HitInfo obtained in TextFlow had its 'text' field set to null, disabling any possibility of computing the index correctly
2. The code that computed the index had an issue in that if an exception happened, it left the object in an inconsistent state, leading to other problems down the road, and more exceptions (incl. AIOB)

What we did here is to compute the insertion index correctly at the source.
I've also checked that in the case of an empty line the insertion index is correct in both cases (at the last line and in the middle of a document).

Reply:
> Assuming line 473 is the one still there today, it looks to me as if that would be reached if you had the caret on an empty line that isn't the last line.

In this case also code hits the condition in 451 as the text run is not null.

>it seems to me that this new code has been pretty much copied from the supposedly buggy code there ..
I'm not sure that code is actually buggy (comments over in that bug), rather that the instance was constructed
with bad data.

Like Andy said, since text was null in `HitInfo::getInsertionIndex()` method in the case of TextFlow, correct insertion index was not calculated. I had to use pretty much same code in PrismTextLayout as I had to consider all the cases especially the U+FE0F Variation Selector-16 characters.

Reply:
I did note that TextFlow passed null which really looked like a WIP to me but its a separate issue.
The exception in 8302511 which is the entire subject of that bug isn't possible if text == null.
My point is that if you pass an incorrect charIndex to BreakIterator.following() you'll get the same exception
in your copy of that code. So how are you making sure you never do that ?

Reply:
The exception in 8302511 was likely caused by a prior exception due to null text, which corrupted insertionIndex value in HitInfo.

The goal of this change is to always compute insertionIndex (an I believe we do it correctly this time).

Reply:
No, it wasn't due to null text. That is provably impossible.

The exception from JDK-8302511 is as follows (copied from the bug report).

Exception in thread "JavaFX Application Thread" java.lang.IllegalArgumentException: offset out of bounds 
at java.base/sun.text.RuleBasedBreakIterator.checkOffset(RuleBasedBreakIterator.java:730) 
at java.base/sun.text.RuleBasedBreakIterator.following(RuleBasedBreakIterator.java:744) 
at javafx.graphics/javafx.scene.text.HitInfo.getInsertionIndex(HitInfo.java:84) 
at javafx.graphics/javafx.scene.text.HitInfo.toString(HitInfo.java:100) 
Look at the code in HitInfo.java referenced in JDK-8302511,  the code at line 84 can only be reached if text != null
..
              if (text != null) {
                    // Skip complex character clusters / ligatures.
                    int next;
                    synchronized(charIterator) {
                        charIterator.setText(text);
                        next = charIterator.following(insertionIndex);   // this is line 84
                    }

...

Reply:
you are right, it cannot be explained by a null text.

so if i understand you correctly, the question is about validity of `insertionIndex += 1` on lines 463 and 469.

463: should it be set to text.length() instead?

469: since text==null, could `insertionIndex = charIndex + 1` be correct?  

And also, is it possible at all that text == null within PrismTextLayout?

Reply:
No, because that's after the call to  BreakIterator.following(int)
It is mainly about the assignment on line 456 
> insertionIndex = charIndex;

since that is what gets passed to BreakIterator.following(int)

and charIndex was calculated by
charIndex = run.getStart() + run.getOffsetAtX(x, trailing);
so it goes back to that calculation.
Perhaps a simple validation that charIndex is not out of bounds is all that is needed.

Reply:
Added check to make sure that insertion index passed to BreakIterator.following(int) is not out of bound.
Added this check after line 456 `insertionIndex = charIndex;`, otherwise insertionIndex remains as -1 and  `Hitinfo::getInsertionIndex` will calculate wrong insertionIndex.
Please check the change.